// @omnedia/ngx-gradient-text/fesm2022/omnedia-ngx-gradient-text.mjs
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';
import * as i0 from '@angular/core';
import { Input, ChangeDetectionStrategy, Component } from '@angular/core';

class NgxGradientTextComponent {
    styleClass;
    set shimmerColor(color) {
        this.style["--gradient-from"] = color;
    }
    set textColor(color) {
        this.style["--gradient-to"] = color;
    }
    text;
    style = {};
    ngOnInit() {
        if (!this.text) {
            throw new Error("om-gradient-text: no text was given to the component!");
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: NgxGradientTextComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "20.0.2", type: NgxGradientTextComponent, isStandalone: true, selector: "om-gradient-text", inputs: { styleClass: "styleClass", shimmerColor: ["gradientStart", "shimmerColor"], textColor: ["gradientEnd", "textColor"], text: "text" }, ngImport: i0, template: "<span class=\"om-gradient-text\" [ngStyle]=\"style\" [ngClass]=\"styleClass\">\r\n  {{ text }}\r\n</span>\r\n", styles: [".om-gradient-text{--gradient-from: #ffaa40;--gradient-to: #9c40ff;color:transparent;-webkit-background-clip:text;background-clip:text;background-size:300% 100%;background-image:linear-gradient(to right,var(--gradient-from),var(--gradient-to),var(--gradient-from));animation:om-gradient 8s linear infinite}@keyframes om-gradient{to{background-position:300% 0}}\n"], dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: NgxGradientTextComponent, decorators: [{
            type: Component,
            args: [{ selector: "om-gradient-text", standalone: true, imports: [CommonModule], changeDetection: ChangeDetectionStrategy.OnPush, template: "<span class=\"om-gradient-text\" [ngStyle]=\"style\" [ngClass]=\"styleClass\">\r\n  {{ text }}\r\n</span>\r\n", styles: [".om-gradient-text{--gradient-from: #ffaa40;--gradient-to: #9c40ff;color:transparent;-webkit-background-clip:text;background-clip:text;background-size:300% 100%;background-image:linear-gradient(to right,var(--gradient-from),var(--gradient-to),var(--gradient-from));animation:om-gradient 8s linear infinite}@keyframes om-gradient{to{background-position:300% 0}}\n"] }]
        }], propDecorators: { styleClass: [{
                type: Input,
                args: ["styleClass"]
            }], shimmerColor: [{
                type: Input,
                args: ["gradientStart"]
            }], textColor: [{
                type: Input,
                args: ["gradientEnd"]
            }], text: [{
                type: Input,
                args: ["text"]
            }] } });

/*
 * Public API Surface of ngx-gradient-text
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgxGradientTextComponent };
